### DISCUSSION TITLE
Trustfall Adapters as WASM Plugins

### DISCUSSION THREAD

**Discussion by ZimboPro** [obi1kenobi/trustfall#564] [Q&A]
So I wrote a tool that allows my team to lint OpenAPI and Terraform [here](https://github.com/ZimboPro/linter) but I want to make it more extensible by making the adapters be WASM plugins (since the ABI is not stable). My concern (lack of knowledge and experience in plugin systems) is how to structure it. So currently, you are only able to pass numbers and strings between the host and the plugin. This is not too bad since we can just serialize the data to a JSON string and pass that. But with the Vertex enum, this might be an issue especially since not all languages will be able to support it.

My other concern is the Adapter/BasicAdapter traits. I am not sure if this _translate_ well across the plugin.

Any ideas, suggestions etc are welcome

**Comment by obi1kenobi** [comment]
Would you be open to hopping on a call to chat about this sometime?

The word "plugin" is one of those that might refer to a few very different things and approaches, and there are a few different ways to think about extensibility as well — e.g. whether you're connecting different adapters to each other, or just supporting multiple adapters in the same tool. I'm worried about misinterpreting your questions and accidentally giving you the wrong answers.

It's currently possible to compile Rust adapters to WASM and run them that way — that's how the rustdoc query playground works (though it's a bit hacky): https://github.com/obi1kenobi/trustfall/tree/main/experiments/trustfall_rustdoc So I think a solution is likely possible, but it is a question of how to best approach the problem to make the solution future-proof.

If you'd be open to hopping on a call, I'd love to chat! What's your schedule like next week? Which days / times would be best?

**Reply by ZimboPro** [reply]
That would be perfect, I am pretty much available any day of the week except Wednesday, Tuesday might be a bit tight though. Time wise, from about 18:00 onwards but my timezone is GMT+2:00, so I am not sure how it compares to your timezone.

**Reply by obi1kenobi** [reply]
I'm on US/Eastern Daylight time (UTC-4). How about Thursday during my afternoon (2pm) and your evening (I believe 8pm): https://time.is/1400_14_Mar_2024_in_New_York ?

If that works, lmk if I should send a calendar invite to the email address on your commits or if you have a different email address you'd prefer it at?
